File: go/emu3.4/instruction.go
package main

type instruction_func func(emu *Emulator)

var instructions [256]instruction_func

func mov_r32_imm32(emu *Emulator) {
	reg := emu.get_code8(0) - 0xB8
	value := emu.get_code32(1)
	emu.registers[reg] = value
	emu.eip += 5
}

func short_jump(emu *Emulator) {
	diff := emu.get_sign_code8(1)
	emu.eip += uint32(diff + 2)
}

func near_jump(emu *Emulator) {
	var diff int32
	diff = emu.get_sign_code32(1)
	emu.eip += uint32(diff + 5)
}

func InitInstructions() {
	for i := 0; i < 8; i++ {
		instructions[0xB8+i] = mov_r32_imm32
	}
	instructions[0xE9] = near_jump
	instructions[0xEB] = short_jump
}

File: rust/emu3-2/src/instruction.rs
use crate::emulator::Emulator;

// mod emulator;

pub const INSTRUCTIONS_COUNT: usize = 256;
pub type Instruction = [Option<fn(&mut Emulator)>; INSTRUCTIONS_COUNT];

pub fn mov_r32_imm32(emu: &mut Emulator) {
    let reg = emu.get_code8(0) - 0xB8;
    let value = emu.get_code32(1);
    emu.set_register32(reg as u32, value);
    emu.eip += 5;
}

pub fn short_jump(emu: &mut Emulator) {
    let value = emu.get_sign_code8(1);
    emu.eip = ((emu.eip as i8) + (value + 2)) as u32;
}

pub fn init_instructions(instructions: &mut Instruction) {
    for i in 0..8 {
        instructions[0xB8 + i] = Some(mov_r32_imm32);
    }
    instructions[0xEB] = Some(short_jump);
}